### ISSUE TITLE
better window create/allocate routines [WIP]

### ISSUE THREAD

**Issue by jeffhammond** [mpiwg-rma/rma-issues#22]
```
int MPI_Win_allocate_new(MPI_Datatype base_datatype,
                         int64_t window_assertions,
                         MPI_Comm comm,
                         void ** baseptr_out,
                         MPI_Win * win)
```

```
int MPI_Win_create_new(const void * baseptr_in,
                       MPI_Datatype base_datatype,
                       int64_t window_assertions,
                       MPI_Comm comm,  
                       MPI_Win * win)
```

# Window assertions

This is a 64-bit field that contains the following entries, which will be implemented as opaque values like other RMA assertions.
```
0 
1

```

**Comment by devreal** [comment]
I like the assertions part and the merger of allocate and allocate-shared.

Just some thoughts on the type safety though:

What are the constraints on the provided datatype? Or what is the semantic of passing an index or vector datatype with non-unit stride? I can see the value of providing contiguous datatypes and the lack of a count argument suggests that allocating more than one element of some base type would require at least a contiguous type. But types with holes? Would we force the implementation to allocate the space or can it allocate space only for the actually described elements?

My other concern is that creating a window may become rather tedious because I have to create the correct datatypes for my window first. Where now I can do

```C
struct A {
  double x, y;
  int z;
};
struct B {
  A a[2]; // expressed either as two distinct members or as contiguous?
  double c;
};

MPI_Win_allocate(..., sizeof(B)*count, ...);
```

I'd now have to create three MPI types up front

```C
MPI_Type_create_struct([DOUBLE, DOUBLE, INT], &A_type);
MPI_Type_create_struct([A_type, A_type, DOUBLE], &B_type);
MPI_Type_create_contiguous(count, B_type, win_type);
MPI_Win_allocate_new(win_type, ...);
```
Changing `A` now requires a change in `A_type` too.

Also, how do I now address `z` in `A`? Are we using byte displacements for communication operations?

**Comment by jdinan** [comment]
The displacement unit is wasted cycles on every RMA operation and should be removed. There's no reasonable way to harmonize displacement units and MPI datatypes. An info assertion regarding the underlying datatypes used on the window would be fine.

**Comment by devreal** [comment]
I'm afraid I don't understand what the benefit of having a datatype attached to a window is in the end. If we end up with byte displacements and arbitrary datatypes in communication calls then why do we want to attach a datatype to the window in the first place?

**Comment by tschuett** [comment]
I really like the assertions for optimisation. Given the datatype upfront is also an optimisation opportunity. Is this going to be a random access one byte window or are going to send complex numbers with two doubles?

**Comment by jeffhammond** [comment]
> I'm afraid I don't understand what the benefit of having a datatype attached to a window is in the end. If we end up with byte displacements and arbitrary datatypes in communication calls then why do we want to attach a datatype to the window in the first place?

Windows with explicit types do not permit byte displacements.  All displacements are in integer multiples of the datatype provided in the window constructor.

**Comment by devreal** [comment]
So only basic types in windows then?

**Comment by jeffhammond** [comment]
...
> Also, how do I now address `z` in `A`? Are we using byte displacements for communication operations?

In that example, you'd use `MPI_BYTE` in the constructor and do explicit displacements.

The point of having types in the constructor are to enable optimization of common cases with homogeneous data.  If you want use RMA to access members of structs that are members of structs, I'm not sure what optimizations you'd expect.

**Comment by jeffhammond** [comment]
> So only basic types in windows then?

No.  If you always access 1000 doubles, you can construct a window that consists of elements of MPI_Datatype corresponding to 1000 doubles.

**Comment by jeffhammond** [comment]
But what I really want to enable is optimization of RMA for arrays of built-in datatypes, since that is an incredibly common use case.

**Comment by devreal** [comment]
> The point of having types in the constructor are to enable optimization of common cases with homogeneous data.

So what is the actual optimization potential here? And how do I make sure that the implementation doesn't make assumptions about me using `MPI_BYTE` when I'm actually using compound types with arbitrary offsets for multi-byte types?

**Comment by jeffhammond** [comment]
> > The point of having types in the constructor are to enable optimization of common cases with homogeneous data.
> 
> So what is the actual optimization potential here? And how do I make sure that the implementation doesn't make assumptions about me using `MPI_BYTE` when I'm actually using compound types with arbitrary offsets for multi-byte types?

If you construct a window with type MPI_DOUBLE, your implementation knows that all RMA accesses are 8-byte aligned and that every accumulate/atomic operation is on `double` and nothing else, which means it can use hardware that supports this type but not other datatypes, which could not otherwise be excluded if the window was allocated using one of the existing calls.

**Comment by jdinan** [comment]
Atomicity already isn't guaranteed across different basic datatypes, right?

**Comment by devreal** [comment]
@jdinan That's right, atomicity is only guaranteed for the same basic types.

@jeffhammond Thanks, that makes sense. That sounds like another hint in the puzzle for atomic performance... What if I know I want to use two types, not just one? Wouldn't it be better to express that information as a list of types in an info key?

**Comment by jdinan** [comment]
I don't see how this would change the atomics implementation. You should be able to use the same approach for `double` regardless of any other datatypes that are in use. What am I missing?

**Comment by devreal** [comment]
Regarding the assertions: I suggest we make it an in/out parameter. The application can ask for features and the implementation will allocate a window even if it cannot satisfy all requirements. It then sets whatever it supports so the application can check whether it can still work with that. If not it has to free the window and move on. I think this simplifies the handshake between both sides if the application can work with less than it requested.

**Comment by jdinan** [comment]
You can also have a flag that says whether the other flags are a hard requirement.

**Comment by jeffhammond** [comment]
I prefer `MPI_Window_assertions input, MPI_Window_assertions &output` and let the user compare `input` and `output`.

**Comment by jeffhammond** [comment]
Notes: add count, full datatype of element memory representation (including non-contiguous struct etc), and atomic requirements as vector of (Op,Type) pairs.

**Comment by devreal** [comment]
From the discussion at the WG meeting today:


1) What is the relation between window creation datatypes and operation datatypes? The first describes the data layout during creation/allocation, the second describes the unit of access for a single operation. We may want to allow for sparse data types and let the implementation ignore the holes. Access displacements would be in bytes, no explicit displacement units handled by the implementation. 
2) We should keep the possibility of using arbitrary data user types (as multiple of `MPI_TYPE_BYTE`) and use atomic operations on other basic datatypes. The creation type should not limit the operations applied to the memory.
3) Add `same_op_noop_replace` to include replacement. Is there any use for mixing different arithmetic/logical operations?

**Comment by devreal** [comment]
Notes from the 9/22/2022 WG meeting:

Extend the interface to provide types and ops intended (in) and supported (out) for accumulate operations:

```C
int MPI_Win_allocate_new(MPI_Aint size, // no displacement, all displacements are in Bytes
                         /* see below */
                         MPI_Window_assertions in_assertions,
                         /* accumulate operations, ignored if user requests support for any op/type combination */
                         int num_types_in,
                         MPI_Datatype types_to_use_in[num_type_ops], 
                         int num_ops_in,
                         MPI_Op ops_to_use_in[num_type_ops],
                         int &num_types_out,
                         MPI_Datatype types_to_use_out[num_type_ops],
                         int &num_ops_out, 
                         MPI_Op ops_to_use_out[num_type_ops],
                         MPI_Count accumulate_max_count, // max number of elements to use in MPI_Accumulate
                         /* communicator */
                         MPI_Comm comm,
                         /* output parameters: MPI signals supported capabilities */
                         MPI_Window_assertions *out_assertions,
                         void ** baseptr_out,
                         MPI_Win * win)
```

Further questions/comments:
- Should we add an info key to for future extension?
- Need to provide `MPI_CAS` to signal the use of CAS as atomic operation. Cannot be used anywhere else.

**Comment by tschuett** [comment]
Please add an info key. There will no `MPI_Win_allocate_new2`.

**Comment by jeffhammond** [comment]
> Need to provide MPI_CAS to signal the use of CAS as atomic operation. Cannot be used anywhere else.

Maybe if we name it `MPI_ATOMIC_CAS`, it's more obvious that isn't not a valid reduce op, and only works with RMA atomics.

**Comment by jeffhammond** [comment]
> Should we add an info key to for future extension?

Yes, every object constructor should have one.

**Comment by jdinan** [comment]
I would like to register a bookmark for #25 which is essentially this ticket, but taken to an extreme. 😁
